### PR TITLE
Reduce configuration time memory overhead

### PR DESCRIPTION
**Pull request by oehme**
I recently tried to profile the configuration time of our own build and found that it was dominated by garbage collection. So I did some allocation profiling and found that we are creating a total of 450MB(!) of garbage. The changes in this pull request are small, transparent fixes which reduce this number by 100MB. Configuration time is about 20% faster thanks to that (down from 2.5s to 2s on my machine).

The fixes are pretty much all independent, so it is best to look at the individual commits. I just combined them on one branch for my own measurement convenience.

### CODE REVIEW

File: subprojects/core/src/main/java/org/gradle/util/Path.java [hunk 1]
```diff
@@ -163,30 +161,37 @@ public Path resolve(String path) {
...
+        int myLength = myPath.length();
+        if (path.startsWith(myPath) && path.length() > myLength + 1 && path.substring(myLength, myLength + 1).equals(Project.PATH_SEPARATOR)) {
+            return path.substring(myLength + 1);
         }
         return path;
     }
 
     private boolean isAbsolutePath(String path) {
-        if (!GUtil.isTrue(path)) {
+        if (path == null || path.isEmpty()) {
```
Comment (bamboo) [inline subprojects/core/src/main/java/org/gradle/util/Path.java]: `Strings.isNullOrEmpty(path)` ?

File: subprojects/core/src/main/java/org/gradle/api/internal/plugins/DefaultConvention.java [hunk 2]
```diff
@@ -322,13 +325,25 @@ public boolean hasMethod(String name, Object... args) {
...
                 if (dynamicObject.hasMethod(name, args)) {
                     return true;
                 }
             }
             return false;
         }
+
+        private BeanDynamicObject asDynamicObject(Object object) {
+            if (dynamicObjects == null) {
+                dynamicObjects = Maps.newHashMap();
```
Comment (bamboo) [inline subprojects/core/src/main/java/org/gradle/api/internal/plugins/DefaultConvention.java]: Maybe an `IdentityHashMap` would make more sense here?

File: build.gradle [hunk 3]
```diff
@@ -220,15 +220,12 @@ allprojects {
...
-                !(otherTask.name in [ 'killExistingDaemons', 'tasks' ])) {
-                otherTask.mustRunAfter cleanTask
-            }
-        }
+    //make sure all tasks run after delete tasks
+    def deleteTasks = tasks.withType(Delete)
+    tasks.matching { task ->
+        !task instanceof Delete && !(task.name in [ 'killExistingDaemons', 'tasks' ])
+    }.all {
+      otherTask.mustRunAfter deleteTasks
```
Comment (bamboo) [inline build.gradle]: The definition of `otherTask` seems to have been lost in the refactoring.

**Comment by oehme** [comment]
@bamboo thanks, I addressed your comments :)

**Comment by bamboo** [comment]
@oehme Good stuff!

**Comment by hansd** [comment]
Sweet!

**Comment by kageiit** [comment]
cool stuff :)

**Pull request merged**